### ISSUE TITLE
default ports

### ISSUE THREAD

**Issue by timoxley** [flowbased/fbp#32]
Having repeated IN/OUT for components that simply have one in and one out seems overly verbose.

Instead of lots of IN/OUT you could simply declare that A -> B -> C would use whatever the 'default' input and output ports are. Could restrict this type of declaration for components with only one in and one out port.

**Comment by timoxley** [comment]
Some related discussion in the comments on https://github.com/noflo/fbp/issues/12

**Comment by jonnor** [comment]
Could allow a top-level declaration of what is the default in ports (cause I think hardcoding in/out is a bit too strict).

```
DEFAULTINPORT=input
DEFAULTOUTPORT=output

a(A) -> b(B) -> c(C)
```
Would connect input to output for a,b,c
in/out could be a "default default" of course.

**Comment by timoxley** [comment]
@jonnor that sounds good to me

**Comment by tlrobinson** [comment]
+1

I like the idea of defaulting the defaults to `in` and `out`, but allowing an override sounds fine too if people want that.

**Comment by tlrobinson** [comment]
Note that this is ambiguous if we don't require `()` after a process name, e.x. does:



```
# Assume Bar and Baz are *both* previously defined processes
In(Input) -> Bar Baz -> Disp(Display)
```

mean

```
In(Input) OUT -> Bar Baz OUT -> IN Disp(Display)
```

or

```
In(Input) OUT -> IN Bar Baz -> IN Disp(Display)
```

???

**Comment by jpaulm** [comment]
Right!  If you want to allow this notation, then you have to require () after process names...  Of course we could go Perlish and have another reserved character, meaning IN or OUT as per context, e.g. 
 
`In(Input) -> Bar Baz  ~ -> Disp(Display)`
`In(Input) -> ~ Bar Baz -> Disp(Display)`
`In(Input) -> ~ Baz  ~ -> Disp(Display)`

**Comment by tlrobinson** [comment]
That's a possibility, but doesn't really seem worth it to me since it actually seems *more* visually cluttered than just writing out the port name.

I assume we can't break backwards compatibility by requiring the parens by default, but maybe we could make it an option (configured by command line or a declaration in the file), which also enables default ports?

I'd love if we could do this in combination with #11, which could make for really concise graphs, almost like shell scripting with the pipe operator:

```
'whatever.txt' -> (ReadFile) -> (Uppercase) -> (Output)
```

**Comment by chadrik** [comment]
> That's a possibility, but doesn't really seem worth it to me since it actually seems more visually cluttered than just writing out the port name.

Agree.

A simple solution is to restrict use of default ports to the case where *both* ports are omitted, instead of just one or the other.  That removes the ambiguity.

**Comment by jpaulm** [comment]
Yes, that's nice!  But we have to make sure the parser reports ambiguous situations, rather than assuming one option or the other...

**Comment by chadrik** [comment]
If you are referring to my suggestion there should not be any ambiguous situations, unless I'm missing something.

**Comment by jpaulm** [comment]
I just meant that, if the developer does leave out a port name by mistake, creating ambiguity, the software should not _assume_ it knows what is meant - it should report the ambiguity. Same applies to the other proposal - if the developer leaves out the ()...

It just occurred to me that, if the developer is designing the application using the DrawFBP diagramming tool, this automatically fills in IN and OUT on each connection, so the developer only needs to change the port names which need to other than IN or OUT.  So this whole debate becomes a bit irrelevant!

**Comment by chadrik** [comment]
> I just meant that, if the developer does leave out a port name by mistake, creating ambiguity, the software should not assume it knows what is meant - it should report the ambiguity.

The parser uses a formal grammar so it is not possible to have that kind of ambiguity: a file with such a mistake would fail to parse.  

For reference, below is the bulk of the grammar:

```
connection 
  = source _ "->" _ connection
  / destination

source
  = bridge
  / outport
  / iip

destination
  = inport
  / bridge

bridge
  = port _ node _ port
  / portWithIndex _ node _ port
  / port _ node _ portWithIndex
  / portWithIndex _ node _ portWithIndex

outport
  = node _ port
  / node _ portWithIndex

iip
  = "'" (iipchar)* "'"

inport 
  = port _ node
  / portWithIndex _ node

node
  = ([a-zA-Z_][a-zA-Z0-9_\-]*) component?

component
  = "(" ([a-zA-Z/\-0-9_]*)? compMeta? ")"

compMeta 
  = ":" [a-zA-Z/=_,0-9]+ 

port
  = basePort __ 

portWithIndex
  = basePort "[" [0-9]+ "]"  __

basePort
  = portname:([a-zA-Z_][a-zA-Z.0-9_]*)

anychar
  = [^\n\r\u2028\u2029]

iipchar
  = [\\][']
  / [^']

_
  = (" "*)?

__
  = " "+
```

You can see that a bridge (e.g. `IN comp OUT`) is defined as:

```
bridge
  = port _ node _ port
  / portWithIndex _ node _ port
  / port _ node _ portWithIndex
  / portWithIndex _ node _ portWithIndex
```

To add the kind of warning you're talking about would first require modifying the grammar to allow it to parse invalid files.  That's a slippery slope, and will probably be difficult to do without other unwanted side-effects.

**Comment by tlrobinson** [comment]
I have implemented this in #58.
